src/proxyvisor.js

```diff
@@ -0,0 +1,1014 @@
+			.select('parentApp')
+			.where({ appId })
+			.then(([{ parentApp }]) => {
+				return this.applications.getTargetApp(parentApp);
+			})
+			.then(parentApp => {
+				return Promise.map(parentApp?.services ?? [], service => {
+					return this.docker.getImageEnv(service.image);
+				}).then(function(imageEnvs) {
+					const imageHookAddresses = _.map(
+						imageEnvs,
+						env =>
+							env.BALENA_DEPENDENT_DEVICES_HOOK_ADDRESS ??
+							env.RESIN_DEPENDENT_DEVICES_HOOK_ADDRESS,
+					);
+					for (const addr of imageHookAddresses) {
+						if (addr != null) {
+							return addr;
+						}
+					}
+					return (
+						parentApp?.config?.BALENA_DEPENDENT_DEVICES_HOOK_ADDRESS ??
+						parentApp?.config?.RESIN_DEPENDENT_DEVICES_HOOK_ADDRESS ??
+						`${constants.proxyvisorHookReceiver}/v1/devices/`
+					);
+				});
+			});
+	}
+
+	sendUpdate(device, timeout, endpoint) {
+		return Promise.resolve(request.getRequestInstance())
+			.then(instance =>
+				instance.putAsync(`${endpoint}${device.uuid}`, {
+					json: true,
+					body: device.target,
+				}),
+			)
+			.timeout(timeout)
+			.spread((response, body) => {
+				if (response.statusCode === 200) {
+					return (this.acknowledgedState[device.uuid] = device.target);
+				} else {
+					this.acknowledgedState[device.uuid] = null;
+					if (response.statusCode !== 202) {
+						throw new Error(`Hook returned ${response.statusCode}: ${body}`);
+					}
+				}
+			})
+			.catch(err => log.error(`Error updating device ${device.uuid}`, err));
+	}
+
+	sendDeleteHook({ uuid }, timeout, endpoint) {
+		return Promise.resolve(request.getRequestInstance())
+			.then(instance => instance.delAsync(`${endpoint}${uuid}`))
+			.timeout(timeout)
+			.spread((response, body) => {
+				if (response.statusCode === 200) {
+					return this.db
+						.models('dependentDevice')
+						.del()
+						.where({ uuid });
+				} else {
+					throw new Error(`Hook returned ${response.statusCode}: ${body}`);
+				}
+			})
+			.catch(err => log.error(`Error deleting device ${uuid}`, err));
+	}
+
+	sendUpdates({ uuid }) {
+		return Promise.join(
+			this.db
+				.models('dependentDevice')
+				.where({ uuid })
+				.select(),
+			this.config.get('apiTimeout'),
+			([dev], apiTimeout) => {
+				if (dev == null) {
+					log.warn(`Trying to send update to non-existent device ${uuid}`);
+					return;
+				}
+				return this.normaliseDependentDeviceFromDB(dev).then(device => {
+					const currentState = formatCurrentAsState(device);
+					const targetState = formatTargetAsState(device);
+					return this.getHookEndpoint(device.appId).then(endpoint => {
+						if (device.markedForDeletion) {
+							return this.sendDeleteHook(device, apiTimeout, endpoint);
+						} else if (
+							device.targetCommit != null &&
+							!_.isEqual(targetState, currentState) &&
+							!_.isEqual(targetState, this.acknowledgedState[device.uuid])
+						) {
+							return this.sendUpdate(device, apiTimeout, endpoint);
```

Comment:
This used to be `@sendUpdate(device, targetState, apiTimeout, endpoint)` but there is no `targetState` accepted so I figured this was the correct fix